### PR TITLE
Boost: Add deactivation survey

### LINKED ISSUE #26921: Add deactivation survey to Jetpack Boost
Discussion: p1666165459396639-slack-C016BBAFHHS

### Summary:
- Similar to the Jetpack plugin show a modal when user clicks the deactivate button on boost
- Provide three options:
  - Cancel -  Cancel the deactivation
  - Deactivate without feedback - Just deactivate the plugin
  - Deactivate with feedback -  Deactivate the plugin and open deactivation survey in a new tab

### PR DESCRIPTION
**Pull request by haqadn**
Fixes #26921

#### Changes proposed in this Pull Request:
* Add a new package to handle deactivation prompts
* Prompt for feedback while deactivating jetpack boost

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
User will be asked to provide their feedback during deactivation

#### Testing instructions:
* Activate Boost
* Go the wp-admin/plugins.php
* Click `deactivate` on Jetpack Boost.
* Repeat the above steps to try all three buttons.

### CODE REVIEW

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/assets/css/deactivation.scss:9]: iirc, nested keyframes are automatically prefixed anyway, so I think it's safe to give this a shorter name.

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/assets/css/deactivation.scss]: The specificity here is quite high.

In addition, I think it would be easier to clearly communicate within the templates that the main/footer elements have special styling significance if they had class names, so that would be a two-fold improvement - fixing specificity and making templates clearer.

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/assets/css/deactivation.scss]: Resets are typically defined at the top

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: Don't add actions to constructors. At the very least, they should be moved to `init`.

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/dialog-template.php]: We should listen for click events from JavaScript instead of adding inline JS.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/class-deactivation-handler.php:119]: I think validation for the dialog view path is necessary somewhere.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: `esc_attr` is for escaping attributes. See https://github.com/Automattic/vip-code-samples/tree/master/10-security#escaping-dynamic-javascript-values for tips on escaping values used in script tags.

**Review by pyronaur — CHANGES_REQUESTED** [review]

File: projects/packages/plugin-deactivation/README.md
Comment (thingalon) [inline projects/packages/plugin-deactivation/README.md]:
```suggestion
Instantiate the `Deactivation_Handler` by providing the `$plugin_slug`, `$feedback_url`, and `$dialog_view_file_path` as parameters.
```

File: projects/packages/plugin-deactivation/README.md
Comment (thingalon) [inline projects/packages/plugin-deactivation/README.md]: 🤔 If you have to provide a `$feedback_url`, then deactivation dialogs using this system will always have a button to provide feedback, a button to cancel, and a button to deactivate without providing feedback.

Why not just build those in, so that people using this module don't have to hook up specific events and provide their own buttons?

**Review by thingalon — CHANGES_REQUESTED** [review]
I love that you built this as a package that other plugins can use.

I have some feedback on the arguments the package takes; IMHO it can be greatly simplified. See inline comments for details. 

However, it works well, and looks great. Happy to use this for Boost. :)

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/assets/css/deactivation.scss:9]: The output isn't adding any additional prefix 
image

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: This is a template and I want the package user to have control over how they want to setup the buttons or if they even want to have the same buttons. For example, jetpack deactivation has links to help resources and missing feedback.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php:119]: What type of validation? Throw an error? How about changing it to a `require`?

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/assets/css/deactivation.scss:9]: Uhoh. Oh well 🤷

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php:119]: `require` is going to break at runtime, meaning it would be more difficult to detect - breaking only on deactivation, and thus - easier to miss.

I'd recommend checking that the file exists when the class is initialized and if the intention is for this to be a fatal error, fatal upon class construction - that way the developer consuming the class is immediately made aware as opposed to sometime later down the road in a specific condition.

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: The same thing effect can be achieved with IDs or class names. You can attach listeners only if your conditions are met.

File: projects/packages/plugin-deactivation/README.md
Comment (haqadn) [reply projects/packages/plugin-deactivation/README.md]: It is probably too early to tell how the dialog will be used. Some may need to have different buttons and have other elements along the buttons. For example jetpack:
image


I am thinking, maybe I should remove the feedback option from the package and let the package consumer handle the functionality. Thoughts on this?

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: It can be. But, it will lock the plugin consumer into having specific buttons. For example, down the road, you want to sever the jetpack connection before deactivating. Having access to those events will let you first run the deactivation and then dispatch the events to control the dialog.


Why don't you like the little bit of inline js though?

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php:119]: I have added a wp_die on `init`

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: > For example, down the road, you want to sever the jetpack connection before deactivating. 

In that case, I might have a script that deals with it and I can call `JetpackPluginDeactivation.events.close` from the same script as well, and if the js selector isn't coupled to style, then it's all nice and clean.


> Why don't you like the little bit of inline js though?

* It adds unnecessary complexity to the code.
* It ties down method names to HTML
  * Long-term maintainability is reduced. For example, if a method name is changed or updated, you'll have to think about all the template files that consumed the "previous behavior" as well. This may lead to a shotgun surgery to make a simple change.
* Looking at the JavaScript file, the flow isn't immediately clear, - you have to know that the events are going to be triggered from outside the script
* It forces the consumer of the package to be responsible for dispatching the events, which makes it more difficult to consume.

Comment (haqadn) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: Having events has the advantage of containing the reaction to a specific plugin.
- If it was a method that needed to be called, then they would also have to pass the plugin slug as a parameter.
- If we attached the events to specific buttons, then we don't give consumers control of the dialog.

Perhaps we can improve the documentation to better explain the flow around events? 

I think both you and Mark(we discussed this on a 1on1 today) have similar concerns about exposing the button functionality even when unnecessary. Mark suggested that we can hide the button-related functionality behind an overridable method. And, I agree. 

However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

**Comment by haqadn** [comment]
Hi @singerb and @oskosk, I am trying to make a package that would handle the plugin deactivation dialog. Would you please take a look and see if the implementation would be easy to consume from the plugins you work with? Do you suggest any changes?

**Comment by singerb** [comment]
@Automattic/yamato-backup @Automattic/yamato-scan FYI.

**Review by thingalon — DISMISSED** [review]
I'm happy for this to go into Boost. We'll need a 👍 from someone with more general approver rights than I have (@oskosk?) to cover the new package you've created.

@singerb I suggested adding you as a reviewer to this, as this kind of thing is likely to be useful for our other standalone plugins, and wanted to get your input. :)

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: I'm not arguing against events. I'm arguing that the listeners should be moved to the class.

> If it was a method that needed to be called, then they would also have to pass the plugin slug as a parameter.

I'm not seeing the problem here. Can't you make it automatically set itself up?

> If we attached the events to specific buttons, then we don't give consumers control of the dialog.

Yes, we do. You control events by moving HTML element IDs around. If you need more control, you can resort to whatever method you prefer for the consumer (including using the current approach). What I'm arguing is that you shouldn't be forced or expected to handle the events when consuming the package.

> However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

This isn't about changing a direction. I'm talking strictly about moving event attachments from one place to another.

Comment (haqadn) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: The listeners are inside the class. What am I missing? https://github.com/Automattic/jetpack/blob/fa13547dacaeb5cffb02ebbfa8022eeb984408e4/projects/packages/plugin-deactivation/src/assets/js/deactivation.ts#L46-L50

Comment (thingalon) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: > However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

I wouldn't hold up for outside reviewers for too long. Most of Jetpack is on a meetup this week. :) 

> The listeners are inside the class. What am I missing?

I thought the same thing, but I think @pyronaur means the `onclick=` listeners. Listening for the click.

Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: You don't need to roll your own event system if you expose open/close methods and either make them overridable or use `wp.hooks` (or both). Then you can set up sensible defaults by searching for certain ID selectors (`jdd-open`, `jdd-close`, `jdd-toggle`, `jdd-etc` ).

Those selectors are easy to move or omit if necessary.

If additional customization is needed, that can be added via a config object or overridable methods or hooks, or all of the above. 

What I'm saying is that you can design a component so that it works out of the box without sacrificing any flexibility.

And to reiterate conversations we've had before - I wouldn't implement the customization part just yet and just focus on providing sensible defaults. If additional options are needed down the road, they can be added via the config object.

To clarify, by config object I mean an object that you pass to your `JetpackPluginDeactivation` upon instantiation, like so:


```js
new JetpackPluginDeactivation('plugin-name, {
  onClose: () => {}
  onOpen: () => {},
  attachEvents: true,
});
```

imho, that config object isn't necessary now, but it can easily be added later if needed without requiring external files like templates to be modified. 

Template files shouldn't be concerned with logic and shouldn't carry the knowledge of what global JavaScript libraries are loaded. `deactivation-dialog.php` is a template file.

File: projects/plugins/boost/app/admin/deactivation-dialog.php
Comment (pyronaur) [inline projects/plugins/boost/app/admin/deactivation-dialog.php:28]: It seems that this button _really_ wants to be an anchor tag.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/assets/js/deactivation.ts]: This is a roundabout way of dealing with events. Adding more events into the mix makes it more difficult to reason about.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/assets/js/deactivation.ts]: I appreciate that there's a separate class name to separate presentation/functionality targets. But I do think that it doesn't have to follow BEM - it's a bit confusing to have a BEM styled class as a JS target.

In the past, I've seen patterns like `js-action-close` class name patterns to indicate that the class has significance only to JavaScript.

But ideally, I think this should probably be an ID selector instead. If you're worried about duplicate IDs on the page if multiple templates are rendered, maybe use the `data-` attribute?

File: projects/packages/plugin-deactivation/README.md
Comment (pyronaur) [inline projects/packages/plugin-deactivation/README.md]: Docs are outdated now I think

**Review by pyronaur — CHANGES_REQUESTED** [review]
We're getting there. But I think we can avoid custom events entirely.

File: projects/plugins/boost/app/admin/deactivation-dialog.php
Comment (haqadn) [reply projects/plugins/boost/app/admin/deactivation-dialog.php:28]: Since we are also navigating to the deactivation link on the same click, the browser detects this as a popup and blocks if we mix up navigation via js and html

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: For this here, I think we need to leverage `wp_localize_script`. I know it's counter intuitive, but `esc_js` doesn't actually escape for script tags, it escapes content for inline javascript attributes: https://developer.wordpress.org/reference/functions/esc_js/

A better solution here would look something like this:

Gather all the plugins that have been registered with `Deactivation_Handler` class in an array, something like:
```php
$plugins = [ 'jetpack', 'jetpack-boost' ]; // just an example of what a populated array would look like, I don't expect anyone to manually hardcode this :) 
```

And then pass those to `jp-plugin-deactivation` using `wp_localize_script` and let WordPress worry about data sanitization.

Then, in `init_dialog`, read those values from where `wp_localize_script` stored them, something like:

```js
window.jetpack_plugin_deactivation_instances = window.jpd_plugin_slug_list.map(
  ( slug ) => new JetpackPluginDeactivation( slug )
)
```

File: projects/plugins/boost/app/admin/deactivation-dialog.php
Comment (pyronaur) [reply projects/plugins/boost/app/admin/deactivation-dialog.php:28]: Oh, that's pretty useful information to add as a comment then.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (pyronaur) [inline projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: I'm not 100% certain this should be a `wp_die`, it definitely needs to be an error visible to a developer, but I'm not sure that this warrants breaking the site.

**Review by pyronaur — CHANGES_REQUESTED** [review]
I think this is on the right track, left a couple more suggestions and `esc_js` usage blocker.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: Instead, would it be appropriate for init to return `true` on success or `WP_Error` on failure? That way, the caller can check for an error, but if it ever ships "broken", it won't cause users to be unable to deactivate.

File: projects/packages/plugin-deactivation/src/assets/css/deactivation.scss
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/assets/css/deactivation.scss]: Current scss is less specificity-ized than this so marking this as resolved.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: This has been done.

File: projects/packages/plugin-deactivation/src/dialog-template.php
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/dialog-template.php]: This is now done via a data attribute on buttons that do things. I like this new solution. Marking this as resolved.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/assets/js/deactivation.ts]: This has been done.

File: projects/packages/plugin-deactivation/README.md
Comment (thingalon) [reply projects/packages/plugin-deactivation/README.md]: This appears to have been updated.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (thingalon) [reply projects/packages/plugin-deactivation/src/assets/js/deactivation.ts]: I don't understand this comment. This seems to be taking all of the elements marked for closing / deactivating, and hooking them up to appropriate callbacks. It seems ok to me?

**Comment by thingalon** [comment]
I've marked conversations that seem to be resolved as done. Honestly, I think this is looking pretty great. I'd just add the requested comment, remove the wp_die, use localize_scripts and let's get this merged. :)

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: Change pushed. It returns the instance on success and `WP_Error` on failure.

File: projects/packages/plugin-deactivation/src/class-deactivation-handler.php
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: It uses `sanitize_key` before coming to `esc_js`. If we used the `wp_localize_script` approach, we would need to run the `init_dialog` function only once. Which has a couple of problems:
1. Consumers cannot control the instantiation individually by overriding the method, which they can do now. Keeping the instantiation separate for each plugin keeps it flexible and simple. e.g. In case we need to pass configuration parameters to customize the behavior as you suggested in one of your earlier comments, we can do that very easily when instantiation is individual.
2. Since it runs only once, the method would probably have to be static.

Comment (thingalon) [reply projects/packages/plugin-deactivation/src/class-deactivation-handler.php]: This seems safe to me.

**Review by thingalon — DISMISSED** [review]
IMHO this is looking great to me :)

**Comment by thingalon** [comment]
Added crew for a review, because this includes a package that is outside my approver ability. :)

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (pyronaur) [reply projects/packages/plugin-deactivation/src/assets/js/deactivation.ts]: But they don't need to be hooked up to callbacks. Instead, `observeDialogActions` could just call the methods like `close()` or `deactivate()` directly without having to dispatch additional events.

**Review by pyronaur — CHANGES_REQUESTED** [review]
`esc_js` shouldn't be used to escape JavaScript within script tags.

**Review by pyronaur — DISMISSED** [review]

**Review by pyronaur — DISMISSED** [review]

**Review by thingalon — DISMISSED** [review]
Looks good, works well. Needs a review from someone with approver access to Jetpack components for the deactivation dialog tool.

File: projects/packages/plugin-deactivation/.gitattributes
Comment (anomiex) [inline projects/packages/plugin-deactivation/.gitattributes]: You'll need to do this for your "dist" directory.

File: projects/packages/plugin-deactivation/package.json
Comment (anomiex) [inline projects/packages/plugin-deactivation/package.json]: Is there a particular reason to use rollup here when almost everything else in the monorepo uses webpack?

The shared monorepo webpack config brings in some nice stuff like script minification and checks for broken JS i18n.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (anomiex) [inline projects/packages/plugin-deactivation/src/assets/js/deactivation.ts:81]: I see in https://github.com/Automattic/jetpack/actions/runs/3390307621/jobs/5634292903#step:8:3800 there's
```
(!) Plugin typescript: @rollup/plugin-typescript TS2339: Property 'href' does not exist on type 'HTMLElement'.
src/assets/js/deactivation.ts: (77:48)
 
77   window.location.href = this.deactivateButton.href;
                                                  ~~~~
```
although for some reason this error doesn't fail the build.

File: projects/packages/plugin-deactivation/.gitattributes
Comment (anomiex) [inline projects/packages/plugin-deactivation/.gitattributes:16]: You'll probably want to add stuff like your `rollup-tsconfig.json` an `rollup.config.js` here so they don't get included in the package or mirror.

File: projects/packages/plugin-deactivation/.gitattributes
Comment (haqadn) [reply projects/packages/plugin-deactivation/.gitattributes]: Thanks, added and changed **dist** to **build** for consistency.

File: projects/packages/plugin-deactivation/package.json
Comment (haqadn) [reply projects/packages/plugin-deactivation/package.json]: There isn't any reason for rollup. I used rollup because we already use it in Boost. But, I agree webpack is better considering the tools. I have replaced rollup with webpack.

File: projects/packages/plugin-deactivation/src/assets/js/deactivation.ts
Comment (haqadn) [reply projects/packages/plugin-deactivation/src/assets/js/deactivation.ts:81]: Fixed

File: projects/packages/plugin-deactivation/.gitattributes
Comment (haqadn) [reply projects/packages/plugin-deactivation/.gitattributes:16]: Now that rollup is replaced with webpack, perhaps webpack should go here? But, I noticed that other packages aren't including webpack in the production-exclude list. I wander if there is a reason I am missing.

Comment (anomiex) [reply projects/packages/plugin-deactivation/.gitattributes:16]: They probably should. The reason is probably just that no one caught it.

**Review by anomiex — COMMENTED** [review]
Other than the linting stuff and the build failure, looks ok to me from a monorepo perspective. I haven't taken a close look at the actual code, I'll trust team review to cover that aspect.

Regarding the build failure, locally it seems to help if I add a `babel.config.js`, along the lines of those used in other packages. That brings in `@babel/preset-typescript` to let Babel recognize typescript syntax, among other things.

**Comment by thingalon** [comment]
pdWQjU-7Y-p2 -- looks like there are some lock errors, taken care of in dfee9a9e9f.

**Review by thingalon — DISMISSED** [review]

**Comment by thingalon** [comment]
@oskosk - would you mind reviewing this? It reaches outside the Boost codebase, and defines a new Jetpack RNA lib to handle deactivation dialogs so my ✅ isn't enough.

File: projects/packages/plugin-deactivation/tsconfig.json
Comment (anomiex) [inline projects/packages/plugin-deactivation/tsconfig.json]: I think the "exclude" is unnecessary, you would only need if there were files in the "include" paths that should be excluded. Since neither of these are, there should be no need to mention them.

File: projects/packages/plugin-deactivation/tsconfig.json
Comment (anomiex) [inline projects/packages/plugin-deactivation/tsconfig.json]: I think you should be good with just
```suggestion
	"include": [ "./src/assets/js" ],
```

**Review by anomiex — DISMISSED** [review]
Looks good from a monorepo structural perspective. I haven't taken a close look at the actual code, I'll trust team review for that part.

I left a few inline suggestions, but I don't think it'll hurt anything if you ignore them. I might even be wrong on them.

**Comment by thingalon** [comment]
Well, this dismissed the reviews on merge 
image

so I may as well tidy up the recommendations. Was going to push them to another P2 so this can finally get merged, but may as well if we have to go through the review dance again.

File: projects/packages/plugin-deactivation/tsconfig.json
Comment (thingalon) [reply projects/packages/plugin-deactivation/tsconfig.json]: Done in 7fd6e02401.

File: projects/packages/plugin-deactivation/tsconfig.json
Comment (thingalon) [reply projects/packages/plugin-deactivation/tsconfig.json]: Done in 7cb96f15b1.

**Review by thingalon — APPROVED** [review]

**Review by anomiex — APPROVED** [review]

**Pull request merged**
